### PR TITLE
Enable debugging of Buildpacks

### PR DESCRIPTION
**Pull request by briandealwis**
**Description**

This PR adds support for Cloud Native Buildpacks for `skaffold debug`.  

At a high level, there are two fundamental changes:
  1. `pkg/skaffold/debug/transform*.go` adds support for configuring a set of _entrypoint launchers_: these are known `ENTRYPOINT`s that effectively launch the CMD value as the command-line, such that the `ENTRYPOINT` can be ignored.  The CNB launcher (`/cnb/lifecycle/launcher`) is such a launcher.  The `node` images also use a `docker-entrypoint.sh` script, which is also an entrypoint launcher. 
  2. `pkg/skaffold/debug/cnb.go` performs some resolution with the CNB image metadata to identify the command-line that will be executed, which is then passed through the normal transformations.

CNB images are normally configured with 1 or more _processes_: each process has a separate _type_ or name, and define a command and arguments.  Each process can also be marked as being _direct_, or whether it needs to be executed as an inline shell script.  If not specified, there is a default process type, `web`. CNB images specifically a launcher (`/cnb/lifecycle/launcher`) to identify and run the right command line.

**User facing changes (remove if N/A)**

`skaffold debug` has experimental support for debugging images produced by Cloud Native Buildpacks.

**Follow-up Work (remove if N/A)**

- Incorporate the node-wrapper from container-debug-support

### CODE REVIEW

File: pkg/skaffold/debug/cnb.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+			var rewriter func(transformed []string) []string
+			if p.Direct {
+				// p.Command is the command and p.Args are the arguments
+				ic.arguments = append([]string{p.Command}, p.Args...)
+				rewriter = func(transformed []string) []string {
+					return append([]string{"--"}, transformed...)
+				}
+			} else {
+				// Script type: split p.Command, pass it through the transformer, and then reassemble in the rewriter.
+				if args, err := shell.Split(p.Command); err == nil {
```
Comment (dgageot) [inline pkg/skaffold/debug/cnb.go:112]: nit: handle the `err != nil` case first

File: pkg/skaffold/debug/cnb.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	if len(ic.arguments) == 0 {
+		// indicates an image mis-configuration as we should have resolved the the
+		// CNB_PROCESS_TYPE (if specified) or `web`.
+		logrus.Warnf("no CNB launch found for %s/%s", ic.artifact, processType)
+		return ic, nil
+	}
+
+	// ic.arguments[0] is a shell script:  split it, pass it through the transformer, and then reassemble in the rewriter.
+	var rewriter func(transformed []string) []string
+	if args, err := shell.Split(ic.arguments[0]); err == nil {
```
Comment (dgageot) [inline pkg/skaffold/debug/cnb.go:135]: What about the error case?

File: pkg/skaffold/debug/cnb.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		remnants := ic.arguments[1:]
+		ic.arguments = args
+		rewriter = func(transformed []string) []string {
+			// reassemble back into a script with arguments
+			return append([]string{shJoin(transformed)}, remnants...)
+		}
+	}
+	return ic, rewriter
```
Comment (dgageot) [inline pkg/skaffold/debug/cnb.go:143]: The rewriter is `nil` here. Is that ok?

File: pkg/skaffold/debug/transform_go.go [hunk 4]
```diff
@@ -61,10 +61,27 @@ func isLaunchingDlv(args []string) bool {
...
 			return true
 		}
 	}
-	if len(config.entrypoint) > 0 {
+
+	// FIXME: as there is currently no way to identify a buildpacks-produced image as holding a Go binary,
+	// nor to cause certain environment variables to be defined in the resulting image, look at the image's
+	// CNB metadata to see if any well-known Go-related buildpacks had been involved.
+	knownGoBuildpackIds := []string{
+		"google.go.runtime", "google.go.build", "google.go.gopath", "google.go.gomod", // GCP Buildpacks
```
Comment (dgageot) [inline pkg/skaffold/debug/transform_go.go:73]: `google.go.build` is always on, all the other on that line are optional ans can't run without `google.go.build`

File: pkg/skaffold/debug/cnb.go (hunk 2 above)
Comment (briandealwis) [reply pkg/skaffold/debug/cnb.go:135]: There's not much we can do, but I'll add a comment.

File: pkg/skaffold/debug/cnb.go (hunk 3 above)
Comment (briandealwis) [reply pkg/skaffold/debug/cnb.go:143]: Yeah, it's checked in the caller.

**Comment by briandealwis** [comment]
PTAL

**Comment by briandealwis** [comment]
PTAL @dgageot

**Comment by briandealwis** [comment]
(Last two commits were to add some debug logging for TestEventsRPC failures.)

**Comment by briandealwis** [comment]
Committed with the event logging in TestEventsRPC as it was useful for catching the errors, and is hidden on successful tests.

**Pull request merged**